XMonad/Actions/EasyMotion.hs

```diff
@@ -0,0 +1,378 @@
+  where
+    rw = max (fi w) (fi th)
+    rh = max (fi h) (fi th)
+
+-- | Create overlay windows the minimum size to contain their key chord
+textSize :: Position -> Rectangle -> Rectangle
+textSize th r = Rectangle { rect_width  = fi th
+                          , rect_height = fi th
+                          , rect_x      = rect_x r + (fi (rect_width r) - fi th) `div` 2
+                          , rect_y      = rect_y r + (fi (rect_height r) - fi th) `div` 2 }
+
+-- | Create overlay windows the full width of the window they select, the minimum height to contain
+--   their chord, and a proportion of the distance from the top of the window they select
+bar :: RealFrac f => f -> Position -> Rectangle -> Rectangle
+bar f th r = Rectangle { rect_width  = rect_width r
+                       , rect_height = fi th
+                       , rect_x      = rect_x r
+                       , rect_y      = rect_y r + round (f' * (fi (rect_height r) - fi th)) }
+                         -- clamp f in [0,1] as other values will appear to lock up xmonad
+                         -- as the overlay will be displayed off-screen
+  where f' = min 0.0 $ max f 1.0
+
+-- | Handles overlay display and window selection. Called after config has been sanitised.
+handleSelectWindow :: EasyMotionConfig -> X (Maybe Window)
+handleSelectWindow EMConf { sKeys = AnyKeys [] } = return Nothing
+handleSelectWindow c = do
+  f <- initXMF $ font c
+  th <- (\(asc, dsc) -> asc + dsc + 2) <$> (textExtentsXMF f (concatMap keysymToString (allKeys . sKeys $ c)))
+  XConf { theRoot = rw, display = dpy } <- ask
+  XState { mapped = mappedWins, windowset = ws } <- get
+  -- build overlays depending on key configuration
+  overlays <- case sKeys c of
+                AnyKeys ks -> buildOverlays ks <$> sortedOverlayWindows
+                  where
+                    visibleWindows :: [Window]
+                    visibleWindows = toList mappedWins
+                    sortedOverlayWindows :: X [OverlayWindow]
+                    sortedOverlayWindows = sortOverlayWindows <$> buildOverlayWindows dpy th visibleWindows
+                PerScreenKeys m -> fmap concat $ sequence $ M.foldr (:) [] $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
+                  where
+                    screenById :: ScreenId -> Maybe (W.Screen WorkspaceId (Layout Window) Window ScreenId ScreenDetail)
+                    screenById sid = L.find ((== sid) . screen) (W.current ws : W.visible ws)
+                    visibleWindowsOnScreen :: ScreenId -> [Window]
+                    visibleWindowsOnScreen sid = L.filter (`elem` (toList mappedWins)) $ W.integrate' $ (screenById sid) >>= W.stack . W.workspace
+                    sortedOverlayWindows :: ScreenId -> X [OverlayWindow]
+                    sortedOverlayWindows sid = sortOverlayWindows <$> (buildOverlayWindows dpy th $ visibleWindowsOnScreen sid)
+    :: X [Overlay]
```

Comment:
With `{-# LANGUAGE ScopedTypeVariables #-}` you can make this type signature much more readable and write `overlays :: X [Overlay] <- case ...` directly.

Reply:
Took a while with the compiler patiently repeating at me:
> `overlays` IS `[Overlay]`, NOT `X [Overlay]`
before I finally listened to it. Hah.

I can't remember if it was this type signature or another one, but I definitely had some frustration fighting correct type signature indentation. And in fact I tried this style to see if it would work, but of course it didn't. `ScopedTypeVariables` is _great_ to know about, thank you.

Reply:
Whoops, sorry about that :)

It's a bit unintuitive that `ScopedTypeVariables` should introduce this behaviour I agree, but it definitely is a very useful extension once you know about it